flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/RowToRowCastRule.java

```diff
@@ -224,4 +224,13 @@ protected String generateCodeBlockInternal(
         writer.stmt(methodCall(writerTerm, "complete")).assignStmt(returnVariable, rowTerm);
         return writer.toString();
     }
+
+    @Override
+    public boolean canFail(LogicalType inputLogicalType, LogicalType targetLogicalType) {
```

Comment:
implement `CollectionToCollectionCastRule` here as well?

Reply:
No, because the logic is not the same: the row to row cast rule allows for target type to have less fields than the source type, so we cannot assert that the children type list has the same size across the two types. Look a couple of lines below when i use `Math.min` and also look at the `matches` method implementation.